libpod/events: Use unix timestamp for event time, and add a timeNano field for event time in nanoseconds

ch33hau opened the pull request:
Fixes: #14993

<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

In case you're only changing docs, make sure to prefix the pull-request title with "[CI:DOCS]". That will prevent functional tests from running and save time and energy.

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?
Yes
<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
The `Time` field in the output of `podman events --format '{{json .}}'` has been changed to `time`, and its value is now a Unix timestamp in seconds. 
Apart from that, a `timeNano` field is also introduced, which is the same as the time field but in nanoseconds precision.
```

rhatdan commented on pkg/domain/entities/events.go:
```diff
@@ -1,11 +1,9 @@
-
 	libpodEvents "github.com/containers/podman/v4/libpod/events"
 	dockerEvents "github.com/docker/docker/api/types/events"
+	"strconv"
```
Please keep the golang native separate from the githup imports.

TomSweeneyRedHat commented:
Changes LGTM once @rhatdan 's comment is addressed and the tests are made happy.

ch33hau commented:
Thanks for reviewing, @rhatdan   @TomSweeneyRedHat !

@rhatdan I think your comment about the `imports` was the reason why `Linter` is unhappy.

I have fixed that linter complaint as well as some bugs (should use `time.Unix(0, nsec)` instead of `time.Unix(sec, 0)`) that caught by `apiv2test`, and also updated a doc. 
It would be nice if there is at least an additional review :) 


By the way, I have a side question which is not directly related to this issue:
Actually, the linter issue is pretty challenging for me to be figured out, if not from your comment, because I am unable to run `make lint` and `make validate` on my machine (Fedora 37, Go 1.19.7).
Running both comments will see the below output on my machine:
```
make validate
VERSION=1.51.1 ./hack/install_golangci.sh
golangci-lint has version 1.51.1 built from b87d2c19 on 2023-02-05T13:06:33Z
Using existing ./bin/golangci-lint has version 1.51.1 built from b87d2c19 on 2023-02-05T13:06:33Z
hack/golangci-lint.sh run

Running golangci-lint for tunnel
Build Tags tunnel: apparmor,seccomp,selinux,remote
Skipped directories tunnel: pkg/api
WARN [linters_context] rowserrcheck is disabled because of generics. You can track the evolution of the generics support by following the https://github.com/golangci/golangci-lint/issues/2649. 
WARN [linters_context] wastedassign is disabled because of generics. You can track the evolution of the generics support by following the https://github.com/golangci/golangci-lint/issues/2649. 
pkg/domain/infra/tunnel/images.go:21:2: "github.com/containers/podman/v4/pkg/errorhandling" imported and not used (typecheck)
	"github.com/containers/podman/v4/pkg/errorhandling"
	^
pkg/domain/infra/tunnel/network.go:12:2: "github.com/containers/podman/v4/pkg/errorhandling" imported and not used (typecheck)
	"github.com/containers/podman/v4/pkg/errorhandling"
	^
pkg/domain/infra/tunnel/pods.go:11:2: "github.com/containers/podman/v4/pkg/errorhandling" imported and not used (typecheck)
	"github.com/containers/podman/v4/pkg/errorhandling"
	^
... more similar errors
```

I will keep investigating how I can make it works but just wanted to ask you if you by chance have any ideas to deal with these problems. (I tried to search from Github issues but seems no other people have similar issues)

Thanks a lot for your time!

ch33hau commented:
Still working on fixing `localsystem` tests.

vrothberg commented:
> This seems to break API?

Yes.  The Event struct is used in the REST API, so we need to keep it stable.

Looking at https://github.com/containers/podman/issues/14993, I am not we can guarantee compatibility on the raw JSON level.  Was there a concrete issue or bug caused by the differences or is it more of a cosmetic nature?

ch33hau commented:
> > This seems to break API?
> 
> Yes. The Event struct is used in the REST API, so we need to keep it stable.
> 
> Looking at #14993, I am not we can guarantee compatibility on the raw JSON level. Was there a concrete issue or bug caused by the differences or is it more of a cosmetic nature?

@edsantiago @vrothberg thanks for mentioning the API changes. That sounds like I should on-hold the changes until a decision has been made, right?

vrothberg commented:
Thanks, @ch33hau.  Yes, let's discuss how to proceed. 

@ch33hau, was there a concrete issue or bug caused by the differences of the raw JSON or is it more of a cosmetic nature?

Luap99 commented:
> Thanks, @ch33hau. Yes, let's discuss how to proceed.
> 
> @ch33hau, was there a concrete issue or bug caused by the differences of the raw JSON or is it more of a cosmetic nature?

The issue links https://github.com/stepchowfun/docuum

Looking at #14993 the docker json output is all lowercase, so technically we could add new fields as lowercase. Converting the timestamp can be done on the client side. Breaking the internal event type must be avoided, this would result in a newer client unable to read the events from the event file.

ch33hau commented:
> Thanks, @ch33hau. Yes, let's discuss how to proceed.
> 
> @ch33hau, was there a concrete issue or bug caused by the differences of the raw JSON or is it more of a cosmetic nature?

@vrothberg Sorry that I am pretty new to Podman so I might not totally get your question correctly. 

What I learned from #14993 was that a user is asking if Podman is trying to match Docker's outputs, because the user is facing some issues when using [a tool](https://github.com/stepchowfun/docuum) that is working fine with Docker but not Podman.

And then [this comment](https://github.com/containers/podman/issues/14993#issuecomment-1193104120) says that `time` and `timeNano` fields should show the Unix timestamp. (at the moment `podman events --format '{{json .}}'` returns the `Time` field that displays a human-readable format and starts with an uppercase, and has no `timeNano` field.)

Luap99 commented:
The unix time can be added to the json but only in the frontend, i.e. cmd/podman/system/events.go. We cannot break our internal structure here, this is also used via our libpod rest API.

You could embed the `event.Time` struct there in a new struct and use that to overwrite some field to include the docker fields as well. In any case we should keep existing fields to not cause regressions for users parsing existing fields.

vrothberg commented:
Friendly ping. @ch33hau, do you have cycles to complete this PR?
